amqp/src/main/scala/akka/stream/contrib/amqp/AmqpSource.scala

```diff
@@ -0,0 +1,153 @@
+        case Some(ex) => failStage(ex)
+        case None     => completeStage()
+      }
+
+      val amqpSourceConsumer = new DefaultConsumer(channel) {
+        override def handleDelivery(consumerTag: String, envelope: Envelope, properties: BasicProperties, body: Array[Byte]): Unit = {
+          consumerCallback.invoke(IncomingMessage(ByteString(body), envelope, properties))
+        }
+
+        override def handleCancel(consumerTag: String): Unit = {
+          // non consumer initiated cancel, for example happens when the queue has been deleted.
+          shutdownCallback.invoke(None)
+        }
+
+        override def handleShutdownSignal(consumerTag: String, sig: ShutdownSignalException): Unit = {
+          // "Called when either the channel or the underlying connection has been shut down."
+          shutdownCallback.invoke(Option(sig))
+        }
+      }
+
+      def setupNamedQueue(settings: NamedQueueSourceSettings): Unit = {
+        channel.basicConsume(
+          settings.queue,
+          false, // never auto-ack
+          settings.consumerTag, // consumer tag
+          settings.noLocal,
+          settings.exclusive,
+          settings.arguments.asJava,
+          amqpSourceConsumer
+        )
+      }
+
+      def setupTemporaryQueue(settings: TemporaryQueueSourceSettings): Unit = {
+        // this is a weird case that required dynamic declaration, the queue name is not known
+        // up front, it is only useful for sources, so that's why it's not placed in the AmqpConnectorLogic
+        val queueName = channel.queueDeclare().getQueue
+        channel.queueBind(queueName, settings.exchange, settings.routingKey.getOrElse(""))
+        channel.basicConsume(
+          queueName,
+          amqpSourceConsumer
+        )
+      }
+
+      settings match {
+        case settings: NamedQueueSourceSettings     => setupNamedQueue(settings)
+        case settings: TemporaryQueueSourceSettings => setupTemporaryQueue(settings)
+      }
+
+    }
+
+    def handleDelivery(message: IncomingMessage): Unit = {
+      if (isAvailable(out)) {
+        pushAndAckMessage(message)
+      } else {
+        if (queue.size + 1 > bufferSize) {
+          failStage(new RuntimeException(s"Reached maximum buffer size $bufferSize"))
+        } else {
+          queue.enqueue(message)
+        }
+      }
+    }
+
+    setHandler(out, new OutHandler {
+      override def onPull(): Unit = {
+        if (queue.nonEmpty) {
+          pushAndAckMessage(queue.dequeue())
+        }
+      }
+
+      override def onDownstreamFinish(): Unit = {
+        // nack any buffered messages to broker if possible
+        if (channel.isOpen) {
+          queue.headOption.foreach(message =>
+            channel.basicNack(message.envelope.getDeliveryTag, true, true))
+        }
+      }
+
+    })
+
+    def pushAndAckMessage(message: IncomingMessage): Unit = {
+      push(out, message)
+      // ack it as soon as we have passed it downstream
```

Comment:
yeah, here a future improvement would be to do something like the kafka offset commits